File: wikipedia/src/main/java/org/wikipedia/beta/views/PlainPasteEditText.java
package org.wikipedia.beta.views;

import android.content.ClipData;
import android.content.ClipDescription;
import android.content.Context;
import android.content.ClipboardManager;
import android.os.Build;
import android.util.AttributeSet;
import android.widget.EditText;

public class PlainPasteEditText extends EditText {
    private Context context;

    public PlainPasteEditText(Context context) {
        super(context);
        this.context = context;
    }

    public PlainPasteEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
        this.context = context;
    }

    public PlainPasteEditText(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        this.context = context;
    }

    @Override
    public boolean onTextContextMenuItem(int id) {
        switch (id) {
            case android.R.id.cut:
                break;
            case android.R.id.paste:
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
                    // Do not allow pasting of formatted text!
                    // We do this by intercepting the clipboard and temporarily replacing its
                    // contents with plain text.
                    ClipboardManager clipboard = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
                    if (clipboard.hasPrimaryClip()) {
                        ClipData oldClipData = clipboard.getPrimaryClip();
                        String lastClipText = oldClipData.getItemAt(oldClipData.getItemCount() - 1).coerceToText(context).toString();
                        // create a new text-only clip data
                        ClipData newClipData = new ClipData(new ClipDescription("text", new String[]{ClipDescription.MIMETYPE_TEXT_PLAIN}),
                                new ClipData.Item(lastClipText));
                        // temporarily set the new clip data as the primary
                        clipboard.setPrimaryClip(newClipData);
                        // execute the paste!
                        boolean ret = super.onTextContextMenuItem(id);
                        // restore the clip data back to the old one.
                        clipboard.setPrimaryClip(oldClipData);
                        return ret;
                    }
                }
                break;
            case android.R.id.copy:
                break;
            default:
        }
        return super.onTextContextMenuItem(id);
    }

}

File: app/src/main/java/org/wikipedia/views/PlainPasteEditText.kt
package org.wikipedia.views

import android.content.ClipData
import android.content.Context
import android.os.Build
import android.text.InputType
import android.util.AttributeSet
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputConnection
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.view.ContentInfoCompat
import androidx.core.view.ViewCompat
import com.google.android.material.textfield.TextInputEditText
import org.wikipedia.R

open class PlainPasteEditText : TextInputEditText {
    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle)

    init {
        // The MIME type(s) need to be set for onReceiveContent() to be called.
        ViewCompat.setOnReceiveContentListener(rootView, arrayOf("text/*"), null)
        // applyCustomCursorDrawable()
    }

    override fun onReceiveContent(payload: ContentInfoCompat): ContentInfoCompat? {
        // Do not allow pasting of formatted text! We do this by replacing the contents of the clip
        // with plain text.
        val clip = payload.clip
        val lastClipText = clip.getItemAt(clip.itemCount - 1).coerceToText(context).toString()
        val updatedPayload = ContentInfoCompat.Builder(payload)
            .setClip(ClipData.newPlainText(null, lastClipText))
            .build()
        return super.onReceiveContent(updatedPayload)
    }

    override fun onCreateInputConnection(outAttrs: EditorInfo): InputConnection? {
        // For multiline EditTexts that specify a done keyboard action, unset the no carriage return
        // flag which otherwise limits the EditText to a single line
        val multilineInput = inputType and InputType.TYPE_TEXT_FLAG_MULTI_LINE == InputType.TYPE_TEXT_FLAG_MULTI_LINE
        val actionDone = outAttrs.imeOptions and EditorInfo.IME_ACTION_DONE == EditorInfo.IME_ACTION_DONE
        if (actionDone && multilineInput) {
            outAttrs.imeOptions = outAttrs.imeOptions and EditorInfo.IME_FLAG_NO_ENTER_ACTION.inv()
        }
        return super.onCreateInputConnection(outAttrs)
    }

    private fun applyCustomCursorDrawable() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            textCursorDrawable = AppCompatResources.getDrawable(context, R.drawable.custom_cursor)
        }
    }
}
